// @pinetwork-js/api-typing/build/routes/v2/payment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cancelPayment = exports.completePayment = exports.verifyTransaction = exports.approvePayment = exports.getPayment = exports.getIncompleteServerPayments = exports.getIncompleteClientPayment = exports.createUserToAppPayment = exports.createAppToUserPayment = void 0;
const routify_1 = require("../routify");
exports.createAppToUserPayment = (0, routify_1.routify) `v2/payments`({});
exports.createUserToAppPayment = (0, routify_1.routify) `v2/payments/user_to_app`({});
exports.getIncompleteClientPayment = (0, routify_1.routify) `v2/payments/incomplete`({});
exports.getIncompleteServerPayments = (0, routify_1.routify) `v2/payments/incomplete_server_payments`({});
exports.getPayment = (0, routify_1.routify) `v2/payments/${'paymentId'}`;
exports.approvePayment = (0, routify_1.routify) `v2/payments/${'paymentId'}/approve`;
exports.verifyTransaction = (0, routify_1.routify) `v2/payments/${'paymentId'}/verify_transaction`;
exports.completePayment = (0, routify_1.routify) `v2/payments/${'paymentId'}/complete`;
exports.cancelPayment = (0, routify_1.routify) `v2/payments/${'paymentId'}/cancel`;
