// @pinetwork-js/sdk/build/message-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageType = void 0;
var MessageType;
(function (MessageType) {
    MessageType["AUTH_ERROR"] = "@pi:app:error:auth";
    MessageType["UNKNOWN_ERROR"] = "@pi:app:error:unknown";
    MessageType["OPEN_APP_CONVERSATION_WITH_ID"] = "@pi:app:conversation:open_with_id";
    MessageType["OPEN_SHARE_DIALOG_ACTION"] = "@pi:app:share_dialog:open";
    MessageType["PREPARE_PAYMENT_FLOW"] = "@pi:app:payments:prepare_payment_flow";
    MessageType["SHOW_PRE_PAYMENT_ERROR"] = "@pi:app:payments:show_pre_payment_error";
    MessageType["START_PAYMENT_FLOW"] = "@pi:app:payments:start_payment_flow";
    MessageType["WAIT_FOR_TRANSACTION"] = "@pi:app:payments:wait_for_transaction";
    MessageType["COMMUNICATION_INFORMATION_REQUEST"] = "@pi:app:sdk:communication_information_request";
    MessageType["OPEN_CONSENT_MODAL"] = "@pi:app:sdk:open_consent_modal";
    MessageType["CHECK_NATIVE_FEATURES"] = "@pi:app:sdk:check_native_features";
    MessageType["DECIDE_CALLBACK_RETRIAL"] = "@pi:app:sdk:decide_callback_retrial";
    MessageType["COPY_TEXT_FROM_TPA"] = "@pi:app:sdk:copy_text_from_tpa";
    MessageType["REQUEST_NATIVE_PERMISSION"] = "@pi:app:sdk:request_native_permission";
    MessageType["OPEN_URL_IN_SYSTEM_BROWSER"] = "@pi:app:sdk:open_url_in_system_browser";
    MessageType["TRACK_LOCATION"] = "@pi:app:sdk:track_location";
})(MessageType || (exports.MessageType = MessageType = {}));
